### PR TITLE
Bootstrap CSS vars

### PR DESCRIPTION
**Pull request by Allcharles**
# Bootstrap CSS variables

Override the bootstrap styles to enable use of CSS variables on most coloured components

## Changes

- Update to bootstrap 5
- Created the following variants for each colour type (`primary`, `secondary`, `info`, etc). Using primary as an example:
  - `--baw-primary-hue`
  - `--baw-primary-saturation`
  - `--baw-primary-lightness`
  - `--baw-primary`
  - `--baw-primary-contrast`
  - `--baw-primary-lighter`
  - `--baw-primary-lighter-contrast`
  - `--baw-primary-lightest`
  - `--baw-primary-lightest-contrast`
  - `--baw-primary-darker`
  - `--baw-primary-darker-contrast`
  - `--baw-primary-darkest`
  - `--baw-primary-darkest-contrast`
- Overrode the styles for the following bootstrap elements
  - buttons
  - nav-links
  - backgrounds
  - text-colors

## Problems

- Would like to have updating of CSS variables possible from an angular service. This would allow for styles to be defined/updated at anytime during runtime or possibly even build time.
- The following colour combinations fail to pass contrast ratios:
  - `--baw-success-lighter` and `--baw-success-lighter-contrast` provides a contrast of 2.74 (needs 4.5)
  - `--baw-info-darker` and `--baw-info-darker-contrast` provides a contrast of 3.06 (needs 4.5)
  - `--baw-warning-darker` and `--baw-warning-darker-contrast` provides a contrast of 2.42 (needs 4.5)

## Issues

Relevant to #1184 

## Visual Changes

image

## Final Checklist

- [x] Assign reviewers if you have permission
- [x] Assign labels if you have permission
- [x] Link issues related to PR
- [x] Ensure project linter is not producing any warnings (`npm run lint`)
- [x] Ensure build is passing on all browsers (`npm run test:all`)
- [x] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

### CODE REVIEW

File: src/app/components/shared/header/header.component.html [hunk 1]
```diff
@@ -1,25 +1,28 @@
-<nav class="container-fluid navbar navbar-expand-lg navbar-dark fixed-top">
+<!--
+  Note: Using [routerLink] instead of [strongRoute] so that routerLinkActive will function.
+  Otherwise uses will not have a visual clue for which page they are on
```
Comment (atruskie) [inline src/app/components/shared/header/header.component.html]: sounds like there's a deficiency in strongRoute? 

File an issue?

File: src/app/styles/_media-mixin.scss [hunk 2]
```diff
@@ -1,10 +1,11 @@
-@import "../../../node_modules/include-media/dist/include-media";
-@import "./variables";
+@import "~include-media/dist/include-media";
```
Comment (atruskie) [inline src/app/styles/_media-mixin.scss:1]: it's unnerving seeing a `~` without a trailing slash. Is this correct?

File: src/app/styles/bootstrap/_root.scss [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  --baw-contrast-threshold: 50%;
+
+  @include create-variable("primary", 216, 98%, 52%);
+  @include create-variable("secondary", 208, 7%, 46%);
+  @include create-variable("success", 152, 69%, 31%);
+  @include create-variable("info", 190, 90%, 50%);
+  @include create-variable("warning", 45, 100%, 51%);
+  @include create-variable("danger", 354, 70%, 54%);
+  @include create-variable("light", 210, 17%, 98%);
+  @include create-variable("dark", 210, 11%, 15%);
```
Comment (atruskie) [inline src/app/styles/bootstrap/_root.scss]: are these all bootstrap defaults? 

maybe we should referencing their original variables?

File: src/app/styles/bootstrap/components/_buttons.scss [hunk 4]
```diff
@@ -0,0 +1,34 @@
+@mixin button($property) {
```
Comment (atruskie) [inline src/app/styles/bootstrap/components/_buttons.scss]: ok, it's unclear to me on a casual read how much is being over-ridden here.

Are we totally replacing the bootstrap `button` mixin?

**Review by atruskie — CHANGES_REQUESTED** [review]
Technically, a great change.

I think anywhere we override something I want to make it plainly clear what is being overridden and why. Document heavily.

I also think, for the sake of clarity, we should namespace non-overridden but custom mixins and functions. 

e.g.

- if we are replacing (overriding) a bootstrap thing then it needs to have the same name
- if we have our own mixing function, prefixed with `baw-` then we can tell easily we're not overriding something

We may roll back some theming changes, but we'll see how it looks after a deploy.

Changes required: documentation of overrides

File: src/app/styles/_media-mixin.scss (hunk 2 above)
Comment (Allcharles) [reply src/app/styles/_media-mixin.scss:1]: yes it is correct

File: src/app/styles/bootstrap/_root.scss (hunk 3 above)
Comment (Allcharles) [reply src/app/styles/bootstrap/_root.scss]: They are bootstrap defaults converted to hsl, AFAIK there is no easy way to convert the rgb values of bootstrap to hsl. I have however just discovered a potential way of handling all of the colour calculations in rgb however if we wanted to use bootstrap values as defaults. I ruled it out as a possibility as you would need to set transparency values on the colours (which would be bad on things like image backgrounds), however it appears i could use the following solution to bypass that: https://gist.github.com/tqc/2564280. Colour contrast calculations may be slightly better/worse as well

File: src/app/styles/bootstrap/components/_buttons.scss (hunk 4 above)
Comment (Allcharles) [reply src/app/styles/bootstrap/components/_buttons.scss]: We are not modifying the bootstrap mixin at all, this has no reference to the way bootstrap sets the button properties, and instead just overrides the end styling. Its effectively just adding styling on top of the bootstrap styling similar to the following:

```css
.btn-primary {
  // bootstrap styling
}

.btn-primary {
  // custom styling
}

File: src/app/components/shared/header/header.component.html (hunk 1 above)
Comment (Allcharles) [reply src/app/components/shared/header/header.component.html]: Tracking issue here: https://github.com/QutEcoacoustics/workbench-client/issues/1216

File: src/app/styles/bootstrap/components/_buttons.scss [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  background-color: unset;
+  color: var(--baw-#{$property});
+}
+
+@each $theme, $default in $theme-colors {
+  .btn-#{$theme} {
+    @include baw-button($theme);
+  }
+
+  .btn-outline-#{$theme} {
```
Comment (atruskie) [inline src/app/styles/bootstrap/components/_buttons.scss:45]: this is the point where the override (extension technically?) actually happens

**Review by atruskie — APPROVED** [review]

File: src/app/styles/bootstrap/components/_buttons.scss (hunk 5 above)
Comment (Allcharles) [reply src/app/styles/bootstrap/components/_buttons.scss:45]: correct

**Pull request merged**